pkg/controller/statefulset/metrics.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package statefulset
+
+import (
+	"sync"
+
+	apps "k8s.io/api/apps/v1"
+	"k8s.io/component-base/metrics"
+)
+
+// StatefulSet is the subsystem name used by this package.
+const StatefulSet = "statefulset"
+
+var (
+	deletedPolicyCounter = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Subsystem:      StatefulSet,
+			Name:           "statefulset_non_retain_when_deleted_policy",
+			Help:           "Number of StatefulSets synced with a non-retain whenDeleted policy.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{},
+	)
+	scaledPolicyCounter = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Subsystem:      StatefulSet,
+			Name:           "statefulset_non_retain_when_scaled_policy",
+			Help:           "Number of StatefulSets synced with a non-retan whenScaled policy.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{},
+	)
+	reconcileSeconds = metrics.NewHistogramVec(
+		&metrics.HistogramOpts{
+			Subsystem:      StatefulSet,
+			Name:           "statefulset_reconcile_seconds",
+			Help:           "StatefulSet reconcile duration, in seconds.",
+			Buckets:        metrics.ExponentialBuckets(1, 1.5, 10),
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{},
+	)
+	unhealthyPodsCounter = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Subsystem:      StatefulSet,
+			Name:           "statefulset_unhealthy_pods_count",
+			Help:           "Count of StatefulSets reconciles that saw an unhealthy pod.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{},
+	)
+)
+
+func recordRetentionPolicyMetrics(set *apps.StatefulSet) {
+	if set == nil {
+		return
+	}
+	policy := set.Spec.PersistentVolumeClaimRetentionPolicy
+	if policy == nil {
+		return
+	}
+	if policy.WhenScaled != apps.RetainPersistentVolumeClaimRetentionPolicyType {
+		scaledPolicyCounter.WithLabelValues().Inc()
```

Comment:
are we explicitly trying to future proof this ? another way to think about this is that when additional retention policy types be introduced, they will automatically be part of this counter

Reply:
Yes, future-proofing is the idea here. As mentioned above, "retain" policy is the default vanilla behavior, so any other policy will cause the metric to increment and presumably be of interest to the cluster operator.

Reply:
See comment https://github.com/kubernetes/kubernetes/pull/111299#discussion_r933627807